### PR TITLE
Feature memory efficency

### PR DESCRIPTION
**Pull request by FFroehlich**
- [x] Reduce memory footprint of `Model` class.
- [x] Add  `rdata-mode` flag that can be set in Solver to 'least-squares'/'gradient'/'full' which limits what is written to rdata.
- [x] Make FIM computable on-demand

### CODE REVIEW

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
FIM currently only available when computing residuals, not likelihoods, which is the opposite of what we want  🙄 , will have to fix.

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

File: src/forwardproblem.cpp [hunk 1]
```diff
@@ -43,8 +43,8 @@ ForwardProblem::ForwardProblem(const ExpData *edata, Model *model,
     }
 }
 
-
 void ForwardProblem::workForwardProblem() {
+    FinalStateStorer ffs(this);
```
Comment (dweindl) [inline src/forwardproblem.cpp:47]:
```suggestion
    FinalStateStorer ffs(this);
```
`fss`?

File: src/model.cpp [hunk 2]
```diff
@@ -754,7 +754,7 @@ void Model::getObservableSensitivity(gsl::span<realtype> sy, const realtype t,
     fdydx(t, x);
     fdydp(t, x);
 
-    this->sx.assign(nx_solver * nplist(), 0.0);
+    this->sx.resize(nx_solver * nplist());
```
Comment (dweindl) [inline src/model.cpp:757]: This does not reset to 0.0 anymore (unless was empty before). Intended?

Comment (FFroehlich) [reply src/model.cpp:757]: Yes. The flatten in the next step makes should make setting to 0.0 redundant and resize is hopefully a bit more efficient.

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

File: include/amici/serialization.h [hunk 3]
```diff
@@ -69,6 +69,7 @@ void serialize(Archive &ar, amici::Solver &u, const unsigned int version) {
     ar &u.ordering;
     ar &u.cpu_time;
     ar &u.cpu_timeB;
+    ar &u.rdata_mode;
```
[the remark this answers is not in the document]
Comment (FFroehlich) [reply include/amici/serialization.h:72]: not sure, we also don't write other status flags like `sensi` and `sensi_meth`

[the remark this answers is not in the document]
Comment (FFroehlich) [reply include/amici/serialization.h:72]: it was indeed missing from write/readSolverData though,

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_swig_interface.py  1
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_swig_interface.py  1
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_swig_interface.py  1
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_pregenerated_models.py  3
- python/tests/test_swig_interface.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Comment by FFroehlich** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- python/tests/test_swig_interface.py  1
- python/tests/test_pregenerated_models.py  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/ICB-DCM/AMICI/pullRequest?prid=5381183&bid=17644352)

**Pull request merged**